src/kvstore/comm_tree.h

```diff
@@ -0,0 +1,518 @@
+            shape_copy[0]   = slice_size;
+            buf.merged.resize(devs_.size());
+            for (unsigned row = 0; row < devs_.size(); ++row) {
+              if (row == devs_.size()-1)
+                shape_copy[0] = last_slice;
+              buf.merged[row] = NDArray(shape_copy, ctx, delay_alloc, type);
+              buf.copy_buf.push_back(std::vector<NDArray>());
+              if (buf.copy_buf[row].empty()) {
+                buf.copy_buf[row].resize(kBranch-1);
+                for (size_t col = 0; col < buf.copy_buf[0].size(); ++col) {
+                  buf.copy_buf[row][col] = NDArray(buf.merged[row].shape(),
+                                                   buf.merged[row].ctx(),
+                                                   delay_alloc,
+                                                   buf.merged[row].dtype());
+                }
+              }
+            }
+          } else {
+            buf.merged.push_back(NDArray(shape, ctx, false, type));
+            if (buf.copy_buf.empty()) {
+              buf.copy_buf.push_back(std::vector<NDArray>());
+              buf.copy_buf[0].resize(kBranch-1);
+              for (size_t col = 0; col < buf.copy_buf[0].size(); ++col) {
+                buf.copy_buf[0][col] = NDArray(buf.merged[0].shape(),
+                                               buf.merged[0].ctx(), delay_alloc,
+                                               buf.merged[0].dtype());
+              }
+            }
+          }
+        }
+      }
+    }
+
+    for (auto it = key_dist.begin(); it != key_dist.end(); ++it) {
+      LOG(WARNING) << "Size " << it->first << " occurs " << it->second << " times";
+    }
+    inited_ = true;
+  }
+
+  std::vector<KeyAttrs> tree_sorted_key_attrs_;
+  /// \brief temporal space for pushing and pulling
+  struct TreeBufferEntry {
+    /// \brief the dense merged value for reduce and broadcast operations
+    std::vector<NDArray> merged;
+    /// \brief the gpu buffer for copy during reduce operation
+    std::vector<std::vector<NDArray>> copy_buf;
+    /// \brief the residual buffer for gradient compression
+    std::vector<NDArray> residual;
+    /// \brief the small buffer for compressed data in sender
+    std::vector<NDArray> compressed_send_buf;
+    /// \brief the small buffer for compressed data in receiver
+    std::vector<NDArray> compressed_recv_buf;
+
+   private:
+    /// \brief the sparse merged value for reduce and rowsparse broadcast operations
+    NDArray sparse_merged;
+  };
+  /// \brief intent of tree_merge_buf_ in old comm.h: store key->gpu mapping
+  ///        new intent: for every gpu: store key->memory mapping
+  std::vector<std::unordered_map<int, TreeBufferEntry>> tree_merge_buf_;
+
+  /// \brief NVLink-connected topology in full binary tree format
+  std::vector<std::vector<size_t>> topology_;
+  std::vector<std::vector<size_t>> scan_;
+  std::vector<Context> devs_;
+
+  /// \brief Highest numbered device
+  int max_dev_;
```

Comment:
I tried to see where this variable is used to ensure that cases when gpus '1,5,3,7' are given work. But it looks like this variable is not used? Please remove this then